File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def _is_same(self, root: Optional[TreeNode], sub_root: Optional[TreeNode]) -> bool:
        if root is None and sub_root is None:
            return True
        if root and sub_root and root.val == sub_root.val:
            return self._is_same(root.left, sub_root.left) and self._is_same(root.right, sub_root.right)
        return False

    def isSubtree(self, root: Optional[TreeNode], sub_root: Optional[TreeNode]) -> bool:
        if sub_root is None:
            return True
        if root is None:
            return False
        if root.val == sub_root.val:
            if self._is_same(root, sub_root):
                return True
        return self.isSubtree(root.left, sub_root) or self.isSubtree(root.right, sub_root)
        

File: Data Structures & Algorithms/invert-a-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private TreeNode recInvertTree(TreeNode root) {
        if (root == null) {
            return null;
        }

        TreeNode tmp = root.left;
        root.left = root.right;
        root.right = tmp;
        
        recInvertTree(root.left);
        recInvertTree(root.right);

        return root;
    }

    // private TreeNode itrInvertTree(TreeNode root) {
    //     Stack<TreeNode> stack = new Stack<>();
    //     TreeNode curr = root;
    //     TreeNode tmp;
    //     while (!stack.empty()) {
            
    //     }
    // }

    public TreeNode invertTree(TreeNode root) {
        return recInvertTree(root);
    }
}
